src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,91 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by, order)
     return sort!(df, _alg, ord)
 end
 
-Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm, o::Base.Sort.Ordering) =
+function Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm,
+                    o::Base.Sort.Ordering)
     permute!(df, _sortperm(df, a, o))
+end
+
+is_complex(o::Ordering) = o isa Union{Order.By, Order.Lt} # Order w/ nonstandard lt or by
+function is_complex(o::UserColOrdering)
+    has_lt = haskey(o.kwargs, :lt)
+    has_by = haskey(o.kwargs, :by)
+    if !has_lt && !has_by
+        return false
+    elseif has_lt && !has_by
+        return o.kwargs[:lt] !== isless
+    elseif has_by && !has_lt
+        return o.kwargs[:by] !== identity
+    elseif has_lt && has_by
+        return o.kwargs[:by] !== identity || o.kwargs[:lt] !== isless
+    end
+end
+
+# Internal function that aids in uniqueness checks
+# Converts column selectors to indices and checks necessary conditions for uniqueness
+function _perform_uniqueness_checks(df::AbstractDataFrame, cols,
+                                    lt::Union{Function, AbstractVector{<:Function}},
+                                    by::Union{Function, AbstractVector{<:Function}},
+                                    order::Union{Ordering, AbstractVector{<:Ordering}})
+
+    if !(lt === isless && by === identity)
+        throw(ArgumentError("Passing either lt or by along with checkunique=" *
+                            "true is not supported."))
+    end
+
+    # Validating the order argument
+    if order isa Ordering
+        order = [order]
+    end
+    for o in order
+        if is_complex(o)
+            throw(ArgumentError("Using either lt or by functions through the " *
+                                "order keyword argument simultaneously with " *
+                                "checkunique=true is not supported."))
+        end
+    end
+
+    # Easiest case, cols contains column indexes already
+    if cols isa AbstractVector{<:ColumnIndex}
+        by_or_lt_set = false
+        col_idxs = cols
+    # Second easiest, multicol index (no vector with orders clauses mixed in)
+    elseif (cols isa MultiColumnIndex && !(cols isa AbstractVector)) || cols isa ColumnIndex
+        by_or_lt_set = false
+        col_idxs = index(df)[cols]
+    elseif cols isa UserColOrdering
+        by_or_lt_set = is_complex(cols)
+        col_idxs = index(df)[_getcol(cols)]
```

Comment:
This is minor, but I assume this is correct, as it should be type stable since `index(df)[_getcol(cols)]` in this case should be an `Int`. Right?

```suggestion
        col_idxs = [index(df)[_getcol(cols)]]
```

Reply:
It should be `Int` since `index` returns the numerical index as far as I understand

Reply:
Yes, so that is why we should wrap it, as `col_idxs` is a vector of `Int` in other branches.